google/cloud/storage/client.h

```diff
@@ -18,9 +18,66 @@
 #include "google/cloud/storage/bucket_metadata.h"
 #include "google/cloud/storage/credentials.h"
 #include "google/cloud/storage/status.h"
+#include "google/cloud/storage/well_known_parameters.h"
 
 namespace storage {
 inline namespace STORAGE_CLIENT_NS {
+class GetBucketMetadataRequest {
+ public:
+  GetBucketMetadataRequest() = default;
+  explicit GetBucketMetadataRequest(std::string bucket_name)
+      : bucket_name_(std::move(bucket_name)) {}
+
+  std::string const& bucket_name() const { return bucket_name_; }
+  GetBucketMetadataRequest& set_bucket_name(std::string bucket_name) {
+    bucket_name_ = std::move(bucket_name);
+    return *this;
+  }
+
+  WellKnownParameters const& well_known_parameters() const {
+    return well_known_parameters_;
+  }
+
+  /**
+   * Apply a list of modifiers to a GetBucketMetadataRequest.
+   *
+   * This is a convenience function to apply multiple modifiers in a single
+   * call. Note that only those modifiers accepted by the `Apply()` overloads
+   * are valid.
+   *
+   * @tparam H the first modifier type
+   * @tparam T the types of the remaining modifiers.
+   * @param head the first modifier in the list.
+   * @param tail the remaining modifiers in the list.
+   */
+  template <typename H, typename... T>
+  void ApplyModifiers(H&& head, T&&... tail) {
+    Apply(std::forward<H>(head));
+    ApplyModifiers(std::forward<T>(tail)...);
+  }
+
+  //@{
+  /// @name Apply a single modifier to the request.
+  void Apply(IfMetaGenerationMatch&& p) {
+    well_known_parameters_.Apply(std::move(p));
```

Comment:
Either add a TODO for this or tack it on now - I don't feel strongly either way - for not allowing users to construct invalid requests via incompatible combinations of preconditions... e.g. you can't set (IfGenerationMatch + IfGenerationNotMatch) or (IfMetagenerationMatch + IfMetagenerationNotMatch). The dotnet lib already does this in its method to modify the request:

https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/59cf05b52d914d687327aadbf0213b5e213ea07d/apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/GetObjectOptions.cs#L76

Reply:
Done.